File: 2364.cpp
// 2364. Count Number of Bad Pairs
#include <bits/stdc++.h>
using namespace std;

// Brute Force Solution
/*
class Solution {
public:
    long long countBadPairs(vector<int>& nums){
        long long badPairs = 0;
        for (int i = 0; i < nums.size(); i++){
            for (int j = i + 1; nums.size(); j++){
                if (i - nums[i] != j - nums[j]){
                    badPairs++;
                }
            }
        }

        return badPairs;
    }
};
*/
/*
class Solution {
public:
    long long countBadPairs(vector<int>& nums){
        long long badPairs = 0;
        map<int, int> diffCount;

        for (int pos = 0; pos < nums.size(); pos++){
            int diff = pos - nums[pos];

            int goodPairsCount = diffCount[diff];

            badPairs += pos - goodPairsCount;

            diffCount[diff] = goodPairsCount + 1;
        }

        return badPairs;
    }
};
*/

class Solution {
public:
    long long countBadPairs(vector<int>& nums){
        long long bad_pairs = 0;
        map<int, int> diff_pair_count;

        for (int i = 0; i < nums.size(); i++){
            int diff = i - nums[i];

            int good_pair_count = diff_pair_count[diff];

            bad_pairs += i - good_pair_count;

            diff_pair_count[diff] = good_pair_count + 1;
        }

        return bad_pairs;
    }
};


int main(){

    return 0;
}

File: 2364.py
# // 2364. Count Number of Bad Pairs
from typing import List

class Solution:
    def countBadPairs(self, nums: List[int]) -> int:
        bad_pairs = 0
        diff_count = {}
        for i in range(len(nums)):
            diff = i - nums[i]

            good_pairs = diff_count.get(diff, 0)

            bad_pairs += i - good_pairs;

            diff_count[diff] = good_pairs + 1
            print(f"diff: {diff}, good_pairs: {good_pairs}, i: {i}, bad_pairs: {bad_pairs}, dff_count: {diff_count}")

        return bad_pairs
    

if __name__ == "__main__":
    s = Solution()
    n = [4, 1, 3, 3]

    result = s.countBadPairs(n)
    print(result)